tests/AbstractProcessTest.php

```diff
@@ -243,6 +243,41 @@ public function testStartAndAllowProcessToExitSuccessfullyUsingEventLoop()
         $this->assertFalse($process->isTerminated());
     }
 
+    public function testProcessWillExitFasterThanExitInterval()
+    {
+        $loop = $this->createLoop();
+        $process = new Process('echo hi');
+        $process->start($loop, 2);
+
+        $time = microtime(true);
+        $loop->run();
+        $time = microtime(true) - $time;
+
+        $this->assertLessThan(0.1, $time);
+    }
+
+    public function testDetectsClosingStdoutWithoutHavingToWaitForExit()
+    {
+        $cmd = 'exec ' . $this->getPhpBinary() . ' -r ' . escapeshellarg('fclose(STDOUT); sleep(1);');
+
+        $loop = $this->createLoop();
+        $process = new Process($cmd);
+        $process->start($loop);
+
+        $closed = false;
+        $process->stdout->on('close', function () use (&$closed) {
+            $closed = true;
+        });
+
+        // run loop for 0.1s only
+        $loop->addTimer(0.1, function () use ($loop) {
+            $loop->stop();
+        });
+        $loop->run();
+
+        $this->assertTrue($closed);
```

Comment:
I'm not sure what this test really tests. The new behavior only changes the `'exit'` event being triggered earlier, no? This test should be unrelated to the exit behavior and probably passed before.

A test that's IMO missing instead is one that tests that closing all pipes from the child _does not_ trigger the `'exit'` event if the process is still running.

Reply:
This test verifies that the "close" event will be emitted without having to wait for the "exit" event.

This PR should not have any effect on this behavior, the test should pass prior to and after applying this patch. Similarly, this PR does not have any effect on the "all pipes closed" behavior as you've suggested.

Reply:
Anyway, I've updated this PR to include some additional tests as suggested :+1: